torch/csrc/autograd/functions/accumulate_grad.h

```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
         // holding references to the Tensor and that is fine since these are not
         // exposed to the user.
         update_grad(new_grad_copy.detach());
+      } else if (
+          !GradMode::is_enabled() && new_grad_copy.is_sparse() &&
+          new_grad_copy._indices().is_contiguous() &&
```

Comment:
We need to add a check to ensure sparse tensors are always created with continguous indicies and values: https://github.com/pytorch/pytorch/issues/12633. I'd like to keep the checks we have here and fix sparse tensor constructor and the logic here in a followup PR.

Reply:
@ezyang Was looking into some of the Sparse Tensor math and it seems like some operations do handle non-contiguous indices/values appropriately: https://bddppq.github.io/codebrowser/pytorch/pytorch/aten/src/ATen/native/sparse/SparseTensorMath.cpp.html#_ZN2at6native18add_out_sparse_cpuERNS_6TensorERKS1_S4_N3c106ScalarE. I'm wondering if its actually fine to build SparseTensors with non-contiguous indices/values?

Reply:
Even if it is fine, I assure you that it is (1) not tested and (2) doesn't work on a bunch of kernels. And unlike noncontiguous dense tensors, where there is a clear value add for non-contiguity (views), there is not really any clear value add to handling these cases correctly. For example, in CUDA:

```
  if (is_same_tensor(r, dense_buffer)) {
    TORCH_CHECK(r_.is_contiguous(), "add: CUDA dense-sparse addition with a non-contiguous output tensor does not work; shout if you need it (see https://github.com/pytorch/pytorch/issues/1521 )");
```

Reply:
It might be possible to "best effort" non-contiguous values/indices by sprinkling contiguous everywhere in sparse but it's not clear to me that this is actually worth it. cc'ing @kurtamohler @nikitaved @hameerabbasi  for any other thoughts.

Reply:
If we're using external libraries in PyTorch, like MKL, those basically demand that we have contiguous values/indices. All literature I've read also implies this indirectly.

Although, of course, we could use non-contiguous versions, that'd be a huge performance hit over the already significant algorithmic hit from using sparse arrays of the same `nnz` as dense arrays. I might be wrong here, but it'd also be bad for GPU sparse arrays.

So, yes, I agree with the sentiment that ensuring contiguousness is a best practice for sparse arrays.